mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -634,6 +639,9 @@ class Parser {
     ast::UserConstraintDecl *exp2Constraint;
     ast::UserConstraintDecl *absConstraint;
   } builtins{};
+
+  // Allows to keep track of included files that were already processed
+  llvm::StringSet<> alreadyIncludedFiles;
```

Comment:
Is this paths to files? If yes, are the absolute or canonicalized? I think you need canonicalized here to check whether files included via different paths are the same. (From what I remember, this is how clang does it)

Reply:
Also, I'd prefer if the name of the variable had a stronger relation to `#once`. It sounds like this tracks all files that were included, but actually it tracks the files that use `#once`.

Reply:
Files are now canonicalized and variable name is strongly linked to `#once`.